// @botonic/react/lib/webchat/devices/device-adapter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeviceAdapter = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _dom = require("../../util/dom");

var _ = require(".");

var _scrollbarController = require("./scrollbar-controller");

var _webchatResizer = require("./webchat-resizer");

var DeviceAdapter = /*#__PURE__*/function () {
  function DeviceAdapter() {
    (0, _classCallCheck2["default"])(this, DeviceAdapter);
  }

  (0, _createClass2["default"])(DeviceAdapter, [{
    key: "init",
    value: function init(host) {
      this.currentDevice = navigator.platform;
      this.webchatResizer = new _webchatResizer.WebchatResizer(this.currentDevice, host);
      this.scrollbarController = new _scrollbarController.ScrollbarController(this.currentDevice, host);
      this.scrollbarController.handleScrollEvents();
    }
  }, {
    key: "onFocus",
    value: function onFocus(host) {
      var _this = this;

      if (this.currentDevice !== _.DEVICES.MOBILE.IPHONE) {
        (0, _dom.scrollToBottom)({
          host: host,
          timeout: 800
        });
        return;
      }

      this.webchatResizer.onFocus(function () {
        return _this.scrollbarController.handleOnTouchMoveEvents();
      });
    }
  }, {
    key: "onBlur",
    value: function onBlur() {
      if (this.currentDevice !== _.DEVICES.MOBILE.IPHONE) return;
      this.webchatResizer.onBlur();
      this.scrollbarController.handleOnTouchMoveEvents();
    }
  }, {
    key: "fontSize",
    value: function fontSize() {
      var defaultFontSize = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 14;
      if (this.currentDevice !== _.DEVICES.MOBILE.IPHONE) return defaultFontSize; // Disabling auto-zoom on input (iPhone devices): https://stackoverflow.com/a/25614477, https://stackoverflow.com/a/6394497

      return 'initial';
    }
  }]);
  return DeviceAdapter;
}();

exports.DeviceAdapter = DeviceAdapter;
