// @botonic/core/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  PROVIDER: true,
  INPUT: true,
  CASE_STATUS: true,
  CASE_RESOLUTION: true,
  Providers: true,
  CoreBot: true,
  cancelHandoff: true,
  deleteUser: true,
  getAgentVacationRanges: true,
  getAvailableAgents: true,
  getAvailableAgentsByQueue: true,
  getOpenQueues: true,
  HandOffBuilder: true,
  humanHandOff: true,
  storeCaseRating: true,
  HubtypeService: true,
  getString: true,
  Router: true,
  isBrowser: true,
  isMobile: true,
  isNode: true,
  params2queryString: true
};
Object.defineProperty(exports, "Providers", {
  enumerable: true,
  get: function get() {
    return _constants.Providers;
  }
});
Object.defineProperty(exports, "CoreBot", {
  enumerable: true,
  get: function get() {
    return _coreBot.CoreBot;
  }
});
Object.defineProperty(exports, "cancelHandoff", {
  enumerable: true,
  get: function get() {
    return _handoff.cancelHandoff;
  }
});
Object.defineProperty(exports, "deleteUser", {
  enumerable: true,
  get: function get() {
    return _handoff.deleteUser;
  }
});
Object.defineProperty(exports, "getAgentVacationRanges", {
  enumerable: true,
  get: function get() {
    return _handoff.getAgentVacationRanges;
  }
});
Object.defineProperty(exports, "getAvailableAgents", {
  enumerable: true,
  get: function get() {
    return _handoff.getAvailableAgents;
  }
});
Object.defineProperty(exports, "getAvailableAgentsByQueue", {
  enumerable: true,
  get: function get() {
    return _handoff.getAvailableAgentsByQueue;
  }
});
Object.defineProperty(exports, "getOpenQueues", {
  enumerable: true,
  get: function get() {
    return _handoff.getOpenQueues;
  }
});
Object.defineProperty(exports, "HandOffBuilder", {
  enumerable: true,
  get: function get() {
    return _handoff.HandOffBuilder;
  }
});
Object.defineProperty(exports, "humanHandOff", {
  enumerable: true,
  get: function get() {
    return _handoff.humanHandOff;
  }
});
Object.defineProperty(exports, "storeCaseRating", {
  enumerable: true,
  get: function get() {
    return _handoff.storeCaseRating;
  }
});
Object.defineProperty(exports, "HubtypeService", {
  enumerable: true,
  get: function get() {
    return _hubtypeService.HubtypeService;
  }
});
Object.defineProperty(exports, "getString", {
  enumerable: true,
  get: function get() {
    return _i18n.getString;
  }
});
Object.defineProperty(exports, "Router", {
  enumerable: true,
  get: function get() {
    return _router.Router;
  }
});
Object.defineProperty(exports, "isBrowser", {
  enumerable: true,
  get: function get() {
    return _utils.isBrowser;
  }
});
Object.defineProperty(exports, "isMobile", {
  enumerable: true,
  get: function get() {
    return _utils.isMobile;
  }
});
Object.defineProperty(exports, "isNode", {
  enumerable: true,
  get: function get() {
    return _utils.isNode;
  }
});
Object.defineProperty(exports, "params2queryString", {
  enumerable: true,
  get: function get() {
    return _utils.params2queryString;
  }
});
exports.CASE_RESOLUTION = exports.CASE_STATUS = exports.INPUT = exports.PROVIDER = void 0;

var _constants = require("./constants");

var _coreBot = require("./core-bot");

var _debug = require("./debug");

Object.keys(_debug).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _debug[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _debug[key];
    }
  });
});

var _handoff = require("./handoff");

var _hubtypeService = require("./hubtype-service");

var _i18n = require("./i18n");

var _router = require("./router");

var _utils = require("./utils");

var PROVIDER = Object.freeze({
  DEV: 'dev',
  FACEBOOK: 'facebook',
  GENERIC: 'generic',
  INTERCOM: 'intercom',
  SMOOCH: 'smooch',
  TELEGRAM: 'telegram',
  TWITTER: 'twitter',
  WEBCHAT: 'webchat',
  WECHAT: 'wechat',
  WHATSAPP: 'whatsapp'
});
exports.PROVIDER = PROVIDER;
var INPUT = Object.freeze({
  TEXT: 'text',
  POSTBACK: 'postback',
  AUDIO: 'audio',
  IMAGE: 'image',
  VIDEO: 'video',
  DOCUMENT: 'document',
  LOCATION: 'location',
  CONTACT: 'contact',
  BUTTON_MESSAGE: 'buttonmessage',
  CAROUSEL: 'carousel',
  CUSTOM: 'custom',
  WEBCHAT_SETTINGS: 'webchatsettings',
  WHATSAPP_TEMPLATE: 'whatsapptemplate'
});
exports.INPUT = INPUT;
var CASE_STATUS = Object.freeze({
  WAITING: 'status_waiting',
  ATTENDING: 'status_attending',
  IDLE: 'status_idle',
  RESOLVED: 'status_resolved'
});
exports.CASE_STATUS = CASE_STATUS;
var CASE_RESOLUTION = Object.freeze({
  OK: 'result_ok',
  NOK: 'result_nok',
  NOT_SOLVED: 'result_not_solved',
  BANNED: 'result_banned'
});
exports.CASE_RESOLUTION = CASE_RESOLUTION;
