// @botonic/react/lib/components/document.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Document = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _core = require("@botonic/core");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _contexts = require("../contexts");

var _message = require("./message");

var _templateObject;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var StyledButton = _styledComponents["default"].a(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: block;\n  height: 25px;\n  background-color: #f0f0f0;\n  padding: 10px;\n  text-align: center;\n  border-radius: 5px;\n  color: black;\n  font-weight: bold;\n  line-height: 25px;\n  text-decoration: none;\n  border: 1px solid black;\n  opacity: 0.9;\n  &:hover {\n    opacity: 1;\n  }\n"])));

var serialize = function serialize(documentProps) {
  return {
    document: documentProps.src
  };
};

var Document = function Document(props) {
  var content = props.children;

  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      getThemeProperty = _useContext.getThemeProperty;

  var documentDownload = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.documentDownload, {});
  if (typeof documentDownload === 'function') documentDownload = documentDownload(props.from);

  if ((0, _core.isBrowser)()) {
    content = /*#__PURE__*/_react["default"].createElement(StyledButton, {
      href: props.src,
      target: "_blank",
      rel: "noreferrer",
      style: _objectSpread({}, documentDownload.style)
    }, documentDownload.text || 'Download');
  }

  return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    role: _constants.ROLES.DOCUMENT_MESSAGE,
    json: serialize(props)
  }, props, {
    type: _core.INPUT.DOCUMENT
  }), content);
};

exports.Document = Document;
Document.serialize = serialize;
