// @botonic/react/lib/components/handoff.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Handoff = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _contexts = require("../contexts");

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

var Container = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  font-family: inherit;\n  flex-direction: column;\n  justify-content: center;\n  align-items: center;\n  padding: 24px;\n"])));

var TransferredContainer = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  text-align: center;\n  white-space: normal;\n"])));

var EndedContainer = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  text-align: center;\n  white-space: normal;\n"])));

var StyledButton = _styledComponents["default"].button(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  max-width: 60%;\n  padding: 12px 24px;\n  background-color: ", ";\n  border: none;\n  border-radius: 4px;\n  margin-top: 8px;\n  cursor: pointer;\n"])), _constants.COLORS.SOLID_WHITE);

var Handoff = function Handoff(props) {
  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      resolveCase = _useContext.resolveCase;

  var _useState = (0, _react.useState)({
    showContinue: true
  }),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      state = _useState2[0],
      setState = _useState2[1];

  var continueClick = function continueClick() {
    setState({
      showContinue: false
    });
    resolveCase();
  };

  var bgColor = state.showContinue ? _constants.COLORS.FRINGY_FLOWER_GREEN : _constants.COLORS.TASMAN_GRAY;
  var fontColor = state.showContinue ? _constants.COLORS.APPLE_GREEN : _constants.COLORS.CACTUS_GREEN;
  return /*#__PURE__*/_react["default"].createElement(Container, {
    style: {
      color: fontColor,
      backgroundColor: bgColor
    }
  }, state.showContinue ? /*#__PURE__*/_react["default"].createElement(TransferredContainer, null, "Conversation transferred to a human agent...") : /*#__PURE__*/_react["default"].createElement(EndedContainer, null, "Human handoff ended"), state.showContinue && /*#__PURE__*/_react["default"].createElement(StyledButton, {
    onClick: continueClick
  }, "Continue"));
};

exports.Handoff = Handoff;
