// @botonic/react/lib/components/image.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Image = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _core = require("@botonic/core");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _message = require("./message");

var _templateObject;

var StyledImage = _styledComponents["default"].img(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  border-radius: 8px;\n  max-width: 150px;\n  max-height: 150px;\n  margin: 10px;\n"])));

var serialize = function serialize(imageProps) {
  return {
    image: imageProps.src
  };
};

var Image = function Image(props) {
  var content = props.children;
  if ((0, _core.isBrowser)()) content = /*#__PURE__*/_react["default"].createElement(StyledImage, {
    src: props.src
  });
  return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    role: _constants.ROLES.IMAGE_MESSAGE,
    json: serialize(props)
  }, props, {
    type: _core.INPUT.IMAGE
  }), content);
};

exports.Image = Image;
Image.serialize = serialize;
