// @botonic/react/lib/components/reply.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Reply = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _constants = require("../constants");

var _contexts = require("../contexts");

var _react2 = require("../util/react");

var _templateObject;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var StyledButton = _styledComponents["default"].button(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  width: 100%;\n  padding: 4px 8px;\n  font-family: inherit;\n  border-radius: 8px;\n  cursor: pointer;\n  outline: 0;\n"])));

var Reply = function Reply(props) {
  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      sendText = _useContext.sendText,
      getThemeProperty = _useContext.getThemeProperty;

  var handleClick = function handleClick(event) {
    event.preventDefault();

    if (props.children) {
      var payload = props.payload;
      if (props.path) payload = "__PATH_PAYLOAD__".concat(props.path);
      sendText(props.children, payload);
    }
  };

  var renderBrowser = function renderBrowser() {
    var replyStyle = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.replyStyle);
    var CustomReply = getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.customReply);

    if (CustomReply) {
      return /*#__PURE__*/_react["default"].createElement("div", {
        onClick: function onClick(e) {
          return handleClick(e);
        }
      }, /*#__PURE__*/_react["default"].createElement(CustomReply, null, props.children));
    }

    return /*#__PURE__*/_react["default"].createElement(StyledButton, {
      style: _objectSpread({
        border: "1px solid ".concat(getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.brandColor, _constants.COLORS.BOTONIC_BLUE)),
        color: getThemeProperty(_constants.WEBCHAT.CUSTOM_PROPERTIES.brandColor, _constants.COLORS.BOTONIC_BLUE)
      }, replyStyle),
      onClick: function onClick(e) {
        return handleClick(e);
      }
    }, props.children);
  };

  var renderNode = function renderNode() {
    if (props.path) {
      var payload = "__PATH_PAYLOAD__".concat(props.path);
      return /*#__PURE__*/_react["default"].createElement("reply", {
        payload: payload
      }, props.children);
    }

    return /*#__PURE__*/_react["default"].createElement("reply", {
      payload: props.payload
    }, props.children);
  };

  return (0, _react2.renderComponent)({
    renderBrowser: renderBrowser,
    renderNode: renderNode
  });
};

exports.Reply = Reply;

Reply.serialize = function (replyProps) {
  var payload = replyProps.payload;
  if (replyProps.path) payload = "__PATH_PAYLOAD__".concat(replyProps.path);
  return {
    reply: {
      title: replyProps.children,
      payload: payload
    }
  };
};
