// @botonic/react/lib/components/text.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Text = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _core = require("@botonic/core");

var _react = _interopRequireWildcard(require("react"));

var _react2 = require("../util/react");

var _markdown = require("./markdown");

var _message = require("./message");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var serializeText = function serializeText(children) {
  children = Array.isArray(children) ? children : [children];
  var text = children.filter(function (e) {
    return !e.type;
  }).map(function (e) {
    if (Array.isArray(e)) return serializeText(e);else return String(e);
  }).join('');
  return text;
};

var serialize = function serialize(textProps) {
  if (!textProps.markdown) return {
    text: serializeText(textProps.children)
  };
  return {
    text: (0, _markdown.serializeMarkdown)(textProps.children)
  };
};
/**
 *
 * @param {TextProps} props
 * @returns {JSX.Element}
 */


var Text = function Text(props) {
  var defaultTextProps = {
    markdown: props.markdown === undefined ? true : props.markdown
  };
  var textProps = (0, _react2.mapObjectNonBooleanValues)(_objectSpread(_objectSpread(_objectSpread({}, props), defaultTextProps), {
    children: _react.Children.toArray(props.children)
  }));
  if (!textProps.markdown) return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    json: serialize(textProps)
  }, textProps, {
    type: _core.INPUT.TEXT
  }), textProps.children);
  return /*#__PURE__*/_react["default"].createElement(_message.Message, (0, _extends2["default"])({
    json: serialize(textProps)
  }, textProps, {
    type: _core.INPUT.TEXT
  }), (0, _markdown.toMarkdownChildren)(textProps.children));
};

exports.Text = Text;
Text.serialize = serialize;
